core/lib/snapshots_applier/src/lib.rs

```diff
@@ -0,0 +1,356 @@
+
+    /// Latency of initial recovery operation split by stage.
+    #[metrics(buckets = Buckets::LATENCIES, unit = Unit::Seconds)]
+    pub initial_stage_duration: Family<InitialStage, Histogram<Duration>>,
+
+    /// Latency of storage log chunk processing split by stage.
+    #[metrics(buckets = Buckets::LATENCIES, unit = Unit::Seconds)]
+    pub storage_logs_chunks_duration: Family<StorageLogsChunksStage, Histogram<Duration>>,
+}
+
+#[vise::register]
+pub(crate) static METRICS: vise::Global<SnapshotsApplierMetrics> = vise::Global::new();
+
+#[derive(thiserror::Error, Debug)]
+pub enum SnapshotsApplierError {
+    #[error("canceled")]
+    Canceled(String),
+    #[error(transparent)]
+    Fatal(#[from] anyhow::Error),
+    #[error("retryable")]
+    Retryable(String),
+}
+pub struct SnapshotsApplier<'a> {
+    connection_pool: &'a ConnectionPool,
+    blob_store: Box<dyn ObjectStore>,
+    applied_snapshot_status: SnapshotRecoveryStatus,
+}
+
+impl From<ObjectStoreError> for SnapshotsApplierError {
+    fn from(error: ObjectStoreError) -> Self {
+        match error {
+            ObjectStoreError::KeyNotFound(_) => SnapshotsApplierError::Fatal(error.into()),
+            ObjectStoreError::Serialization(_) => SnapshotsApplierError::Fatal(error.into()),
+            ObjectStoreError::Other(_) => SnapshotsApplierError::Retryable(
+                "An error occurred while accessing object store".to_string(),
+            ),
+        }
+    }
+}
+
+impl From<SqlxError> for SnapshotsApplierError {
+    fn from(err: SqlxError) -> Self {
+        match err {
+            SqlxError::Database(_)
+            | SqlxError::RowNotFound
+            | SqlxError::ColumnNotFound(_)
+            | SqlxError::Configuration(_)
+            | SqlxError::TypeNotFound { type_name: _ } => SnapshotsApplierError::Fatal(err.into()),
+            _ => SnapshotsApplierError::Retryable(format!(
+                "An error occured when accessing DB: {err}"
+            )),
+        }
+    }
+}
+
+impl From<RpcError> for SnapshotsApplierError {
+    fn from(err: RpcError) -> Self {
+        match err {
+            Error::Transport(_) | Error::RequestTimeout | Error::RestartNeeded(_) => {
+                SnapshotsApplierError::Retryable(
+                    "An error occurred while accessing main node".to_string(),
+                )
+            }
+            _ => SnapshotsApplierError::Fatal(err.into()),
+        }
+    }
+}
+
+#[async_trait]
+pub trait SnapshotsApplierMainNodeClient: fmt::Debug + Send + Sync {
+    async fn fetch_l2_block(&self, number: MiniblockNumber) -> Result<Option<SyncBlock>, RpcError>;
+
+    async fn fetch_newest_snapshot(&self) -> Result<Option<SnapshotHeader>, RpcError>;
+}
+impl<'a> SnapshotsApplier<'a> {
+    pub async fn load_snapshot(
+        connection_pool: &'a ConnectionPool,
+        main_node_client: Box<dyn SnapshotsApplierMainNodeClient>,
+        blob_store: Box<dyn ObjectStore>,
```

Comment:
Nit: Inconsistency among arguments: `connection_pool` is supplied by ref, while others are supplied by value. I'd suggest to supply them all by ref.